pilot/pkg/dns/dns.go

```diff
@@ -0,0 +1,302 @@
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package dns
+
+import (
+	"net"
+	"strings"
+	"sync/atomic"
+	"time"
+
+	"github.com/miekg/dns"
+
+	"istio.io/pkg/log"
+
+	nds "istio.io/istio/pilot/pkg/proto"
+)
+
+// Holds configurations for the DNS downstreamServer in Istio Agent
+type LocalDNSServer struct {
+	// Holds the pointer to the DNS lookup table
+	lookupTable atomic.Value
+
+	downstreamMux    *dns.ServeMux
+	downstreamServer *dns.Server
+
+	// This is the upstreamClient used to make upstream DNS queries
+	// in case the data is not in our cache.
+	upstreamClient    *dns.Client
+	resolvConfServers []string
+	// The namespace where the proxy resides
+	// determines the hosts used for shortname resolution
+	proxyNamespace string
+	// Optimizations to save space and time
+	proxyDomain      string
+	proxyDomainParts []string
+}
+
+// Borrowed from https://github.com/coredns/coredns/blob/master/plugin/hosts/hostsfile.go
+type DNSLookupTable struct {
+	// The key is a FQDN matching a DNS query (like example.com.), the value is pre-created DNS RR records
+	// of A or AAAA type as appropriate.
+	name4 map[string][]dns.RR
+	name6 map[string][]dns.RR
+}
+
+const (
+	defaultTTL = 3600 // 1hr
+)
+
+func NewLocalDNSServer(proxyNamespace, proxyDomain string) (*LocalDNSServer, error) {
+	h := &LocalDNSServer{
+		downstreamMux:  dns.NewServeMux(),
+		proxyNamespace: proxyNamespace,
+	}
+
+	// proxyDomain could contain the namespace making it redundant.
+	// we just need the .svc.cluster.local piece
+	parts := strings.Split(proxyDomain, ".")
+	if len(parts) > 0 {
+		if parts[0] == proxyNamespace {
+			parts = parts[1:]
+		}
+		h.proxyDomainParts = parts
+		h.proxyDomain = strings.Join(parts, ".")
+	}
+
+	h.downstreamMux.Handle(".", h)
+
+	h.upstreamClient = &dns.Client{
+		Net:         "udp",
+		DialTimeout: 3 * time.Second,
+		ReadTimeout: 500 * time.Millisecond,
+	}
+
+	// We will use the local resolv.conf for resolving unknown names.
+	dnsConfig, err := dns.ClientConfigFromFile("/etc/resolv.conf")
+	if err != nil {
+		log.Warnf("failed to load /etc/resolv.conf: %v", err)
+		return nil, err
+	}
+
+	// Unlike traditional DNS resolvers, we do not need to append the search
+	// namespace to a given query and try to resolve it. This is because the
+	// agent acts as a DNS interceptor for DNS queries made by the application.
+	// The application's resolver is already sending us DNS queries, one for each
+	// of the DNS search namespaces. We simply need to check the existence of this
+	// name in our local nametable. If not, we will forward the query to the
+	// upstream resolvers as is.
+	//
+	// NOTE: There is still a NxN issue here as the app will cycle through N
```

Comment:
not sure I understand the NxN? I see the `N` issue, but not n^2?

Reply:
Aren't we not going through the full cycle in agent?

Reply:
Agent is but any thing that intercepts DNS transparently will potentially undergo NxN 
where app tries to resolve name across all nameservers. And for each query from the app, the agent will try to do the same across all nameservers. In most situations this wont be an issue because there is typically only one nameserver. In any case, this is way better than envoy because envoy used to also search through all the DNS search namespaces.  so if the resolv.conf had 5 search namespaces, the total number of dns queries would be 25 (5 from app * 5 from envoy) to arrive at a result that says the host is not found. We dont have this problem in the agent because we dont do the search namespace stuff (as we control the dns implementation). So it ends up being just 5 in total (if num name servers = 1).